File: polyglot/typescript/scan_directory.ts
import { promises as fs } from 'fs';
import { join, readdir, existsSync } from 'path';

interface Rule {
  id: string;
  text: string;
  regex: boolean;
}

interface Match {
  ruleId: string;
  filePath: string;
  content: string;
  matches: Array<{ offset: number; length: number }>;
}

async function scanDirectory(directoryPath: string, rules: Rule[]): Promise<Match[]> {
  const matches: Match[] = [];

  // Check if the directory exists
  if (!existsSync(directoryPath)) {
    throw new Error(`Directory not found: ${directoryPath}`);
  }

  // Read all files in the directory
  const files = await readdir(directoryPath, { withFileTypes: true });

  for (const file of files) {
    if (file.isDirectory()) continue; // Skip directories

    const filePath = join(directoryPath, file.name);
    const content = await fs.readFile(filePath, 'utf-8');

    // Check each rule against the file content
    for (const rule of rules) {
      const regex = new RegExp(rule.text, 'g');
      let match;
      const fileMatches: Array<{ offset: number; length: number }> = [];

      while ((match = regex.exec(content)) !== null) {
        fileMatches.push({
          offset: match.index,
          length: match[0].length
        });
      }

      if (fileMatches.length > 0) {
        matches.push({
          ruleId: rule.id,
          filePath,
          content,
          matches: fileMatches
        });
      }
    }
  }

  return matches;
}

// Example usage
async function runDemo() {
  const directoryPath = './test-data'; // Replace with your directory path
  const rules: Rule[] = [
    {
      id: 'rule1',
      text: 'secret_key',
      regex: false
    },
    {
      id: 'rule2',
      text: '\\b[A-Z0-9]{8}\\b',
      regex: true
    }
  ];

  try {
    const results = await scanDirectory(directoryPath, rules);
    console.log('Matches found:');
    for (const match of results) {
      console.log(`Rule: ${match.ruleId}`);
      console.log(`File: ${match.filePath}`);
      console.log(`Matches: ${match.matches.map(m => `Offset: ${m.offset}, Length: ${m.length}`).join(', ')}`);
      console.log();
    }
  } catch (error) {
    console.error('Error scanning directory:', error);
  }
}

// Run the demo
runDemo();

File: polyglot/python/scan_directory.py
import os
import re
from typing import List, Dict, Any

class YaraRule:
    def __init__(self, name: str, pattern: str, description: str = ""):
        self.name = name
        self.pattern = pattern
        self.description = description

    def match(self, content: str) -> bool:
        return re.search(self.pattern, content, re.DOTALL) is not None

def load_rules_from_file(file_path: str) -> List[YaraRule]:
    rules = []
    with open(file_path, 'r') as f:
        rule_data = ""
        for line in f:
            line = line.strip()
            if line.startswith("rule"):
                if rule_data:
                    rules.append(YaraRule(**eval(rule_data)))
                rule_data = line
            elif line.startswith("description") or line.startswith("strings") or line.startswith("condition"):
                rule_data += "\n" + line
        if rule_data:
            rules.append(YaraRule(**eval(rule_data)))
    return rules

def scan_directory(directory: str, rules: List[YaraRule]) -> Dict[str, List[Dict[str, Any]]]:
    results = {}
    for root, _, files in os.walk(directory):
        for file in files:
            file_path = os.path.join(root, file)
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
            except UnicodeDecodeError:
                continue
            matches = []
            for rule in rules:
                if rule.match(content):
                    matches.append({
                        "rule": rule.name,
                        "description": rule.description,
                        "file": file_path
                    })
            if matches:
                results[file_path] = matches
    return results

def main():
    rules_file = 'rules.yar'
    if not os.path.exists(rules_file):
        print(f"Rules file '{rules_file}' not found.")
        return

    rules = load_rules_from_file(rules_file)
    directory_to_scan = os.getcwd()
    scan_results = scan_directory(directory_to_scan, rules)

    if scan_results:
        print("Matches found:")
        for file_path, matches in scan_results.items():
            print(f"\nFile: {file_path}")
            for match in matches:
                print(f" - Rule: {match['rule']}, Description: {match['description']}")
    else:
        print("No matches found.")

if __name__ == "__main__":
    main()